text/0015-interest-management.md

```diff
@@ -0,0 +1,261 @@
+* Feature Name: `interest-management`
+* Start Date: 2020-05-11
+* RFC PR: TBA
+* Issue: TBA0
+
+# Summary
+[summary]: #summary
+This RFC proposes an interest management system to help MLAPI support greater scale in games.  It is sometimes conceptualized as an 'AOI' (Area of Interest) system, but goes beyond that in also providing support for later elements such as prioritization.
+
+# Motivation
+[motivation]: #motivation
+
+In the existing MLAPI Interest system, every networked object (i.e. `NetworkedObject`) stores a list of all clients that can see that object (in the variable `observers`).  In the default case, every object stores a list of all connected clients.  This creates some scalability problems
+
+* Every client has to be checked against every object
+* Storage: every object has to store a list of all clients
+* Updating: every time a client joins / leaves, every object must be scanned and modified
+
+For example, a leading FPS game begins with around 100 clients and 50,000 replicated objects.  With the current scheme, one would need to store 20 MB of client / object data and do 50,000*100 = 5 million client/object checks each tick.
+
+
+## Desired Outcome
+
+What we instead would like to achieve:
+
+* We want to minimize the computational per-frame work.  As the library walks through its connected clients, we want a system that will efficiently answer the question 'which objects should this connection see'
+* Therefore, we desire an interface that takes in a connection and returns the objects for that client
+* We want the way this question is answered to be user-extensible to support all kinds of different scenarios.  Domain decomposition / geometric / AOI is very common, but we also want developers to be able to come up with creative, game-specific **schemes** that have nothing to do with spatial relationships
+* We want the aforementioned schemes to be composable.  For instance, a user should be able to have both a AOI scheme and a "always synchronize players who are in the same cohort" scheme and have the results from each efficiently combine
+* Each scheme should be free to decide how to implement relationships; so long as the scheme can take in a client and return a set of objects it should be free to come up with that answer using whatever storage and computation it wants
+* Each scheme can therefore decide at what cadence it does its work.  A scheme can choose to dynamically come up with an answer every frame.  Or, it could return cached results each time except every 30th frame.
```

Comment:
Is the "every 30th frame" pertaining to the tick update?  What consumes every 30th frame?